### PR TITLE
VexRiscv integration

### PR DESCRIPTION
**Pull request by Dolu1990**
Hi,

There the VexRiscv integration, it was tested with the simple.py target with iverilog and on the DE1-SoC.
Everything look fine (bios + serial).

Just notice that the reset vector of the CPU is fixed in the https://github.com/m-labs/VexRiscv-verilog project.

There is the GCC installation commands (from https://github.com/SpinalHDL/VexRiscv#build-the-risc-v-gcc) : 

```sh
wget https://static.dev.sifive.com/dev-tools/riscv64-unknown-elf-gcc-20171231-x86_64-linux-centos6.tar.gz
tar -xzvf riscv64-unknown-elf-gcc-20171231-x86_64-linux-centos6.tar.gz
riscv64-unknown-elf-gcc-20171231-x86_64-linux-centos6.tar.gz
sudo mv riscv64-unknown-elf-gcc-20171231-x86_64-linux-centos6 /opt/riscv64-unknown-elf-gcc-20171231-x86_64-linux-centos6
sudo mv /opt/riscv64-unknown-elf-gcc-20171231-x86_64-linux-centos6 /opt/riscv
echo 'export PATH=/opt/riscv/bin:$PATH' >> ~/.bashrc 
```

On the DE1-SoC, the synthesis  tool can go up to 133 Mhz. So with an Artix7, it should be capable to go at 200+ Mhz. If it doesn't, It is likely that some pipelining need to be added on the instruction/data bus. It is simple to do, and would not add noticable performance penality. Let's me know if it appear to be required in larger SoC.

Regards
Charles

### CODE REVIEW

**Comment by mithro** [comment]
@cr1901 @enjoy-digital - FYI - You might be very interested in this pull request.

**Comment by sbourdeauducq** [comment]
Some quick comments (more related to VexRiscv):
* use synchronous reset
* can the reset vector be configurable, e.g. by making it an input port that then gets tied to a constant? We do use two CPUs with different reset vectors in ARTIQ.

**Comment by sbourdeauducq** [comment]
Also VexRiscv-verilog does not have to know about or mention misoc (that can probably simplify the needlessly long path src/main/scala/misoc/cores/vexriscv).

File: misoc/cores/vexriscv/core.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+from misoc.interconnect import wishbone
+
+
+class VexRiscv(Module):
+    def __init__(self, platform):
+        self.ibus = i = wishbone.Interface()
+        self.dbus = d = wishbone.Interface()
+        self.interrupt = Signal(32)
+
+        self.timerInterrupt = Signal(1)
```
Comment (sbourdeauducq) [inline misoc/cores/vexriscv/core.py]: Please use PEP8 names, i.e. ``self.timer_interrupt``.

File: misoc/integration/builder.py [hunk 2]
```diff
@@ -169,7 +169,17 @@ def _initialize_rom(self):
...
+                        # Little endian swap
+                        w = w[::-1]
                     boot_data.append(struct.unpack(">I", w)[0])
+
+            if self.soc.cpu_type == "vexriscv":
+                # rom CRC endian reswap
+                buf = 0
+                for b in range(4):
+                    buf |= ((boot_data[-1] >> 8*b) & 0xFF) << (24-8*b)
+                boot_data[-1] = buf
```
Comment (sbourdeauducq) [inline misoc/integration/builder.py]: Can't this be done externally e.g. by objcopy?

**Comment by cr1901** [comment]
@mithro When this PR is merged, I'll make an equivalent PR for `litex` if @Dolu1990 has not done so already.

File: misoc/integration/builder.py [hunk 3]
```diff
@@ -169,7 +169,17 @@ def _initialize_rom(self):
                     w = boot_file.read(4)
                     if not w:
                         break
+                    if self.soc.cpu_type == "vexriscv":
+                        # Little endian swap
+                        w = w[::-1]
                     boot_data.append(struct.unpack(">I", w)[0])
```
Comment (sbourdeauducq) [inline misoc/integration/builder.py]: If it *really* has to be done here, ``struct.unpack`` can convert endianness already, just use ``<I``.

File: misoc/software/bios/boot-helper-vexriscv.S [hunk 4]
```diff
@@ -0,0 +1,4 @@
+.section    .text, "ax", @progbits
+.global     boot_helper
+boot_helper:
+    nop
```
Comment (sbourdeauducq) [inline misoc/software/bios/boot-helper-vexriscv.S]: EOL

File: misoc/software/libbase/system.c [hunk 5]
```diff
@@ -62,6 +71,13 @@ void flush_cpu_dcache(void)
 
 	for (i = 0; i < cache_size; i += cache_block_size)
 		mtspr(SPR_DCBIR, i);
+#elif defined (__vexriscv__)
+    unsigned long cacheInfo = read_csr(0xCC0);
```
Comment (sbourdeauducq) [inline misoc/software/libbase/system.c]: ``cache_info`` etc.

**Comment by sbourdeauducq** [comment]
> can the reset vector be configurable, e.g. by making it an input port that then gets tied to a constant? We do use two CPUs with different reset vectors in ARTIQ.

Additionally different boards use different reset vectors (due to different memory layouts in the flash used for XIP). So, we really need that.

File: misoc/integration/cpu_interface.py [hunk 6]
```diff
@@ -12,6 +12,10 @@ def get_cpu_mak(cpu):
         triple = "or1k-linux"
         cpuflags = "-mhard-mul -mhard-div -mror -mffl1 -maddc"
         clang = "1"
+    if cpu == "vexriscv":
```
Comment (sbourdeauducq) [inline misoc/integration/cpu_interface.py]: ``elif``
This breaks all other CPUs as written.

File: misoc/integration/cpu_interface.py [hunk 7]
```diff
@@ -12,6 +12,10 @@ def get_cpu_mak(cpu):
...
+        triple = "riscv64-unknown-elf" # GCC from https://static.dev.sifive.com/dev-tools/riscv64-unknown-elf-gcc-20171231-x86_64-linux-centos6.tar.gz
```
Comment (sbourdeauducq) [inline misoc/integration/cpu_interface.py]: Toolchain installations instructions should be in the doc, not the source.

**Comment by sbourdeauducq** [comment]
Uses only 321 LUTs more than LM32 on Kasli - looking good.

**Comment by sbourdeauducq** [comment]
```
/home/sb/misoc/misoc/software/libbase/system.c: In function 'flush_l2_cache':
/home/sb/misoc/misoc/software/libbase/system.c:100:3: error: output operand constraint lacks '='
   __asm__ volatile("lw x0, 0(%0)\n":"r"(addr));
   ^~~~~~~
/home/sb/misoc/misoc/software/libbase/system.c:91:24: warning: unused variable 'dummy' [-Wunused-variable]
  register unsigned int dummy;
                        ^~~~~
/home/sb/misoc/misoc/software/libbase/system.c:100:3: error: output operand constraint lacks '='
   __asm__ volatile("lw x0, 0(%0)\n":"r"(addr));
   ^~~~~~~
/home/sb/misoc/misoc/software/libbase/system.c:100:3: error: invalid lvalue in asm output 0
Makefile:19: recipe for target 'system.o' failed
make: *** [system.o] Error 1
```

**Comment by sbourdeauducq** [comment]
sdram.c needs an update as well (use nop in cdelay).

**Comment by sbourdeauducq** [comment]
We got liftoff!
```
$ flterm --port /dev/ttyUSB_kasli-1_0
[FLTERM] Starting...

MiSoC BIOS
(c) Copyright 2007-2017 M-Labs Limited
Built Apr 24 2018 11:49:00

BIOS CRC passed (0edfc341)
Initializing SDRAM...
Read delays: 1:15-24  0:15-25  completed
Memtest OK
Booting from serial...
Press Q or ESC to abort boot completely.
sL5DdSMmkekro
Timeout
No boot medium found
BIOS> mr 0 256
Memory dump:
0x00000000  6f 00 00 0b 13 00 00 00 13 00 00 00 13 00 00 00  o...............
0x00000010  13 00 00 00 13 00 00 00 13 00 00 00 13 00 00 00  ................
0x00000020  23 2e 11 fe 23 2c 51 fe 23 2a 61 fe 23 28 71 fe  #...#,Q.#*a.#(q.
0x00000030  23 26 a1 fe 23 24 b1 fe 23 22 c1 fe 23 20 d1 fe  #&..#$..#"..# ..
0x00000040  23 2e e1 fc 23 2c f1 fc 23 2a 01 fd 23 28 11 fd  #...#,..#*..#(..
0x00000050  23 26 c1 fd 23 24 d1 fd 23 22 e1 fd 23 20 f1 fd  #&..#$..#"..# ..
0x00000060  13 01 01 fc ef 00 80 08 83 20 c1 03 83 22 81 03  ......... ..."..
0x00000070  03 23 41 03 83 23 01 03 03 25 c1 02 83 25 81 02  .#A..#...%...%..
0x00000080  03 26 41 02 83 26 01 02 03 27 c1 01 83 27 81 01  .&A..&...'...'..
0x00000090  03 28 41 01 83 28 01 01 03 2e c1 00 83 2e 81 00  .(A..(..........
0x000000a0  03 2f 41 00 83 2f 01 00 13 01 01 04 73 00 20 30  ./A../......s. 0
0x000000b0  17 11 00 10 13 01 01 f5 17 05 00 10 13 05 85 f4  ................
0x000000c0  97 05 00 10 93 85 c5 05 63 08 b5 00 23 20 05 00  ........c...# ..
0x000000d0  13 05 45 00 6f f0 5f ff 37 15 00 00 13 05 05 88  ..E.o._.7.......
0x000000e0  73 10 45 30 ef 00 90 4b 6f 00 00 00 f3 27 00 36  s.E0...Ko....'.6
0x000000f0  73 27 00 33 b3 77 f7 00 93 f7 17 00 63 84 07 00  s'.3.w......c...
BIOS> 
```
Apart from the few details I mentioned, looks really nice, congratulations.

**Comment by jbqubit** [comment]
Wow. That was quick from PR to run! Cool!  

@Dolu1990 What are prospects for IEEE Floating point in VexRiscv? Wikipedia says this already part of the RISC-V v2.0 ISA.

File: misoc/integration/builder.py (hunk 2 above)
Comment (whitequark) [reply misoc/integration/builder.py]: objcopy doesn't have anything for this, and in any case I don't see how bringing in *more* external dependencies makes anything better. This implementation is fine.

File: misoc/integration/cpu_interface.py [hunk 8]
```diff
@@ -12,6 +12,10 @@ def get_cpu_mak(cpu):
...
+        cpuflags = "-D__vexriscv__ -march=rv32im  -mabi=ilp32 -nostdlib -lgcc -mcmodel=medany -nostartfiles -ffreestanding -Wl,-Bstatic"
```
Comment (whitequark) [inline misoc/integration/cpu_interface.py]: All these flags for freestanding compilation should go to the Makefile (and most of them are already there). `-lgcc` shouldn't appear at all, I think. `-Wl,-Bstatic` is likely unnecessary.

File: misoc/software/include/base/irq.h [hunk 9]
```diff
@@ -9,6 +9,39 @@ extern "C" {
...
+    asm volatile ("csrrs %0, " #reg ", %1" : "=r"(__tmp) : "r"(bit)); \
+  __tmp; })
+
+#define clear_csr(reg, bit) ({ unsigned long __tmp; \
+  if (__builtin_constant_p(bit) && (unsigned long)(bit) < 32) \
+    asm volatile ("csrrc %0, " #reg ", %1" : "=r"(__tmp) : "i"(bit)); \
+  else \
+    asm volatile ("csrrc %0, " #reg ", %1" : "=r"(__tmp) : "r"(bit)); \
+  __tmp; })
+#endif
```
Comment (whitequark) [inline misoc/software/include/base/irq.h]: All these should go to `system.h`, and named the same as their instructions, just like in or1k.

File: misoc/software/libbase/crt0-vexriscv.S [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+  la a0, _fbss
+  la a1, _ebss
+bss_loop:
+  beq a0,a1,bss_done
+  sw zero,0(a0)
+  add a0,a0,4
+  j bss_loop
+bss_done:
+
+  li a0, 0x880     //880 enable timer + external interrupts
```
Comment (whitequark) [inline misoc/software/libbase/crt0-vexriscv.S]: Does this enable interrupts? If so this should not happen, since other misoc targets do not enter main with interrupts enabled.

File: misoc/software/include/base/irq.h [hunk 11]
```diff
@@ -44,6 +81,8 @@ static inline unsigned int irq_getmask(void)
 	return mask;
 #elif defined (__or1k__)
 	return mfspr(SPR_PICMR);
+#elif defined (__vexriscv__)
+    return read_csr(0x330);
```
Comment (whitequark) [inline misoc/software/include/base/irq.h]: We shouldn't have magic values, add a header with register definitions instead.

File: misoc/software/libbase/system.c [hunk 12]
```diff
@@ -80,6 +96,8 @@ void flush_l2_cache(void)
 		__asm__ volatile("lw %0, (%1+0)\n":"=r"(dummy):"r"(addr));
 #elif defined (__or1k__)
 		__asm__ volatile("l.lwz %0, 0(%1)\n":"=r"(dummy):"r"(addr));
+#elif defined (__vexriscv__)
+		__asm__ volatile("lw x0, 0(%0)\n":"r"(addr));
```
Comment (whitequark) [inline misoc/software/libbase/system.c]: We could simply replace this chunk of assembly with a volatile load.

File: misoc/software/libbase/system.c [hunk 13]
```diff
@@ -62,6 +71,13 @@ void flush_cpu_dcache(void)
 
 	for (i = 0; i < cache_size; i += cache_block_size)
 		mtspr(SPR_DCBIR, i);
+#elif defined (__vexriscv__)
+    unsigned long cacheInfo = read_csr(0xCC0);
+    unsigned long cacheWaySize = cacheInfo & 0xFFFFF;
+    unsigned long cacheLineSize = (cacheInfo >> 20) & 0xFFF;
+    for(register unsigned long idx asm ("a0") = 0;idx < cacheWaySize;idx += cacheLineSize){
```
Comment (whitequark) [inline misoc/software/libbase/system.c]: Register-pinned variables are not supported in clang, so rewrite this to move from a compiler-provided register to `a0`.

Also, doesn't the assembler support a mnemonic for this instruction? Or is this a custom one of vexriscv?

File: .gitignore [hunk 14]
```diff
@@ -3,3 +3,7 @@ __pycache__
 *.egg-info
 *.vcd
 outgoing
+build
+dist
+.idea
```
[the remark this answers is not in the document]
Comment (Dolu1990) [reply .gitignore]: the .idea is made by pycharm (the python IDE from jetbrains)

[the remark this answers is not in the document]
Comment (whitequark) [reply .gitignore]: You should put it into your global .gitignore (`~/.gitignore`)

File: misoc/software/libbase/crt0-vexriscv.S (hunk 10 above)
Comment (Dolu1990) [reply misoc/software/libbase/crt0-vexriscv.S]: The comment isn't explicit enough.
There is another layer of interrupt enable flag (CSR mstatus.MIE) in addition of those timer/external interrupt enable which enable/disable interrupt.
So everything is fine :)

File: misoc/software/include/base/irq.h (hunk 11 above)
Comment (Dolu1990) [reply misoc/software/include/base/irq.h]: system.h too ?

Comment (whitequark) [reply misoc/software/include/base/irq.h]: Sure

Comment (whitequark) [reply misoc/software/include/base/irq.h]: Hm no, not system.h, there is `spr-defs.h` for or1k, so `csr-defs.h` for riscv should make sense.

Comment (Dolu1990) [reply misoc/software/include/base/irq.h]: Unfortunately : 
read_csr(DEFINED_CONSTANT) 
-> Error: Instruction csrr requires absolute expression

Also, can't do as the or1k do, as the read_csr macro require compilation time constants.

File: misoc/software/libbase/system.c (hunk 12 above)
Comment (Dolu1990) [reply misoc/software/libbase/system.c]: I agree, but hadn't touch it as it looked like black magic ^^

File: misoc/software/include/base/irq.h (hunk 11 above)
Comment (whitequark) [reply misoc/software/include/base/irq.h]: You can just leave out the __builtin_constant_p branch. The miniscule reduction in performance doesn't matter.

File: misoc/software/libbase/system.c (hunk 13 above)
Comment (Dolu1990) [reply misoc/software/libbase/system.c]: What i can do is implementing it as a assembly function.

This is a custom instruction.
The RISC-V spec doesn't specify cache flush/clean instruction (On their linux capable cores they implement memory coherency)

Comment (whitequark) [reply misoc/software/libbase/system.c]: Just  use `asm volatile (".word ..." :: "{a0}"(idx));` which should put the index in place

File: misoc/software/libbase/system.c (hunk 12 above)
Comment (whitequark) [reply misoc/software/libbase/system.c]: It's just touching enough code pages that everything else in cache gets evicted.

File: .gitignore (hunk 14 above)
[the remark this answers is not in the document]
Comment (Dolu1990) [reply .gitignore]: Thanks, wasn't aware about this feature.

File: misoc/software/libbase/system.c (hunk 13 above)
Comment (Dolu1990) [reply misoc/software/libbase/system.c]: asm volatile(".word(0b01110000000001010101000000001111)" :: "x10"(idx));
         ^~~
/home/spinalvm/hdl/misoc/misoc/software/libbase/system.c:79:9: error: matching constraint references invalid operand number

Comment (whitequark) [reply misoc/software/libbase/system.c]: Ok, then `asm volatile("mov x10, %0 \n .word(0b01110000000001010101000000001111)"::"r"(idx))`

Comment (Dolu1990) [reply misoc/software/libbase/system.c]: Great, that work-around is fine, thanks.

File: misoc/software/include/base/irq.h (hunk 11 above)
Comment (Dolu1990) [reply misoc/software/include/base/irq.h]: Unfortunatly, we can't do the same way than or1k, as RISC-V do not provide a instruction to access CSR pointed by a CPU register.

The CSR address realy need to be known at compilation time.
=> asm volatile ("csrr %0, %1" : "=r"(mask) : "i"(CSR_IRQ_MASK));   + csr-defs.h

Comment (whitequark) [reply misoc/software/include/base/irq.h]: oh WTF gross. Sure, keep them as macros, then, I guess, no other choice...

@jordens One more reason to not have mtspr/mfspr in ARTIQ Python.

File: misoc/integration/builder.py [hunk 15]
```diff
@@ -165,11 +165,19 @@ def _initialize_rom(self):
...
                 boot_data = []
+                unpackEndian = ">I" if self.soc.cpu_type != "vexriscv" else "<I"
                 while True:
                     w = boot_file.read(4)
                     if not w:
                         break
-                    boot_data.append(struct.unpack(">I", w)[0])
+                    boot_data.append(struct.unpack(unpackEndian, w)[0])
+
+            if self.soc.cpu_type == "vexriscv":
```
Comment (whitequark) [inline misoc/integration/builder.py]: You've already swapped the endianness above, no need to do it again.

File: misoc/integration/cpu_interface.py [hunk 16]
```diff
@@ -12,6 +12,10 @@ def get_cpu_mak(cpu):
         triple = "or1k-linux"
         cpuflags = "-mhard-mul -mhard-div -mror -mffl1 -maddc"
         clang = "1"
+    elif cpu == "vexriscv":
+        triple = "riscv64-unknown-elf"
+        cpuflags = "-D__vexriscv__ -march=rv32im  -mabi=ilp32 -mcmodel=medany -nostartfiles"
```
Comment (whitequark) [inline misoc/integration/cpu_interface.py]: `-nostartfiles` is also inappropriate here.

File: misoc/software/include/base/irq.h [hunk 17]
```diff
@@ -9,6 +9,11 @@ extern "C" {
 #include <system.h>
 #endif
 
+#ifdef __vexriscv__
+#include <system.h>
+#include <csr-defs.h>
+#endif
```
Comment (whitequark) [inline misoc/software/include/base/irq.h]: I think `system.h` should be included unconditionally.

File: misoc/software/include/base/irq.h [hunk 18]
```diff
@@ -31,6 +38,8 @@ static inline void irq_setie(unsigned int ie)
 		mtspr(SPR_SR, mfspr(SPR_SR) | SPR_SR_IEE);
 	else
 		mtspr(SPR_SR, mfspr(SPR_SR) & ~SPR_SR_IEE);
+#elif defined (__vexriscv__)
+    if(ie) csrs(mstatus,CSR_MSTATUS_MIE); else csrc(mstatus,CSR_MSTATUS_MIE);
```
Comment (whitequark) [inline misoc/software/include/base/irq.h]: Please fix indentation here and elsewhere. The C code in misoc uses tabs.

File: misoc/integration/builder.py (hunk 15 above)
Comment (Dolu1990) [reply misoc/integration/builder.py]: You mean the "rom CRC endian reswap" is useless ?
Because in simulation, without that reswap, the CRC check fail.

File: misoc/integration/builder.py (hunk 2 above)
Comment (sbourdeauducq) [reply misoc/integration/builder.py]: We already use objcopy here. I'm surprised that it can only produce an image with the wrong endianness.

File: misoc/integration/builder.py (hunk 15 above)
Comment (whitequark) [reply misoc/integration/builder.py]: Why are you swapping everything twice for vexriscv? There is some bug here. I don't know where but it exists.

Comment (Dolu1990) [reply misoc/integration/builder.py]: I can only see one swap, for a single 32 bit word (the one of the CRC). 

The first stuff with the struct.unpack isn't there to change the endianess, but only to decode the file and generate the rom words.
FileBytes[0] -> rom[0][7:0]
FileBytes[1] -> rom[0][15:8]
FileBytes[2] -> rom[0][24:16]
FileBytes[3] -> rom[0][31:24]
FileBytes[4] -> rom[1][7:0]

n.b. RISC-V is little endian, LM32 is big endian

Comment (whitequark) [reply misoc/integration/builder.py]: Oh sorry, I misread.

Comment (Dolu1990) [reply misoc/integration/builder.py]: The thing which trouble me the most, is that between the LM32 and the VexRiscv (Big vs Little endian) i had nothing to change to have the serial peripheral working ^.^

Apparently both agree that LSB is bit zero and MSB is bit 31 (on the 32 bit whishbone bus).

File: misoc/integration/builder.py (hunk 2 above)
Comment (sbourdeauducq) [reply misoc/integration/builder.py]: Oh right; then I think it's better to change the CRC generation function to add an option to select endianness.

File: misoc/integration/builder.py (hunk 15 above)
Comment (sbourdeauducq) [reply misoc/integration/builder.py]: Yes? what is wrong with that?

**Comment by Dolu1990** [comment]
@jbqubit About floating point, sure, the RISC-V specification + compiler stuff fully support it. 

Currently in VexRiscv nothing is implemented about it. Currently i'm focussing to get the support of the compressed instruction set, then maybe some rework on the datacache, and they why not the floating point stuff : D

File: misoc/cores/vexriscv/core.py [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+
+class VexRiscv(Module):
+    def __init__(self, platform, cpu_reset_address):
+        self.ibus = i = wishbone.Interface()
+        self.dbus = d = wishbone.Interface()
+
+        self.interrupt = Signal(32)
+        self.timer_interrupt = Signal(1)
+
+        self.external_reset_vector = Signal(32)
```
Comment (sbourdeauducq) [inline misoc/cores/vexriscv/core.py]: This doesn't need to have a signal, and doesn't need to be a property of the object. Just do ``i_externalResetVector=cpu_reset_address``.

File: misoc/integration/builder.py (hunk 2 above)
Comment (sbourdeauducq) [reply misoc/integration/builder.py]: @Dolu1990 Can you get to the bottom of the various endianness issues and propose minimal, clean code?

File: misoc/software/common.mak [hunk 20]
```diff
@@ -59,6 +63,9 @@ export CFLAGS_$(subst -,_,$(CARGO_TRIPLE)) = $(CFLAGS)
 
 # Linker options
 LDFLAGS = --gc-sections -nostdlib -nodefaultlibs -L$(BUILDINC_DIRECTORY)
+ifeq ($(CPU),vexriscv)
+	LDFLAGS += -nostartfiles
```
Comment (sbourdeauducq) [inline misoc/software/common.mak]: Why is this needed for risc-v but not the other CPUs?

File: .gitignore [hunk 21]
```diff
@@ -2,4 +2,4 @@ __pycache__
 *.pyc
 *.egg-info
 *.vcd
-outgoing
+outgoing
```
Comment (sbourdeauducq) [inline .gitignore]: EOL

File: misoc/integration/builder.py (hunk 2 above)
Comment (sbourdeauducq) [reply misoc/integration/builder.py]: BIOS CRC is added here:
https://github.com/m-labs/misoc/blob/master/misoc/software/bios/Makefile
https://github.com/m-labs/misoc/blob/master/misoc/tools/mkmscimg.py

File: misoc/software/libbase/crt0-vexriscv.S [hunk 22]
```diff
@@ -0,0 +1,77 @@
...
+  sw zero,0(a0)
+  add a0,a0,4
+  j bss_loop
+bss_done:
+
+  li a0, 0x880  //880 enable timer + external interrupt sources (until mstatus.MIE is set, they will never trigger an interrupt)
+  csrw mie,a0
+
+  call main
+infinitLoop:
```
Comment (sbourdeauducq) [inline misoc/software/libbase/crt0-vexriscv.S]: Can we avoid javaNames here as well?

File: misoc/software/common.mak (hunk 20 above)
Comment (Dolu1990) [reply misoc/software/common.mak]: From what i have seen for RISC-V, when you want do to baremetal, you add -nostartfiles  -nostdlib  as compiler flags. 
After some try, it look like that if you are not using the std libs, it will not try to add a startup file. But it's better to not play with the fire. 

I have no idea for other CPU. It probably depend how far they implemented the clib and stuff like this. But i'm not a c guru. Maybe i'm wrong.

**Comment by Dolu1990** [comment]
Everything should be fine now, let's me know if i skipped something

File: misoc/software/common.mak (hunk 20 above)
Comment (sbourdeauducq) [reply misoc/software/common.mak]: I don't think this can really go wrong in obscure ways... if it still works without ``-nostartfiles`` I think those three lines should be removed.

Comment (whitequark) [reply misoc/software/common.mak]: `-nodefaultlibs` implies `-nostartfiles`.

**Comment by whitequark** [comment]
Looks good to me after fixing the `-nostartfiles` part.

File: misoc/software/common.mak (hunk 20 above)
Comment (Dolu1990) [reply misoc/software/common.mak]: From https://gcc.gnu.org/onlinedocs/gcc/Link-Options.html

-nodefaultlibs
 ... The standard startup files are used normally, unless -nostartfiles is used.

But -nostdlib implies -nostartfiles (from the same link)

So great, i removed it :D
